Hide non-interactive elements on multi-spectator players

frenzibyte opened the pull request:
Closes #14261 

Video to quickly showcase how this looks now:

https://user-images.githubusercontent.com/22781491/129304090-f55d4d4d-55bf-41b5-9818-47ab36772054.mp4

frenzibyte committed:
Hide player settings overlay on multi-spectator player loader

frenzibyte committed:
Hide and disable player settings overlay on multi-spectator player

frenzibyte committed:
Add test coverage

smoogipoo commented:
What about the hold to quit button?

peppy commented:
We should be removing all the interactive elements that can't be clicked (until they can) I guess?

frenzibyte commented:
Yeah seems the "hold to quit" button is the only uninteractive elements which should be hidden, the song progress should remain to point out the current time of the spectator player, but the bar should be hidden like in regular solo mode. Will do that.

smoogipoo commented:
Fwiw this wasn't the way I was going to go about this - I was thinking of making `HUDOverlay` derivable instead and a custom `MultiSpectatorPlayerHUDOverlay` or something like that. Because it's a bit haphazard with the way that the "Watching X" text is added to `SpectatorPlayer` itself completely outside of the HUD...

But hiding individual elements seems fine as a stop-gap solution.

peppy commented:
Honestly as this is probably a temporary thing, I'd just call `.Expire()` on the elements, keeping all the logic in one place? Not sure..

frenzibyte commented:
Expiring does seem to work and simplify logic, have switched to it now.

frenzibyte committed:
Expire instead of hiding and disabling visibility state

Since it's a temporary change until the spectator interface gets
improved, no need to add further logic.

frenzibyte committed:
Expire hold-to-quit button on multi-spectator player

frenzibyte committed:
Hide song progress bar on multi-spectator player

frenzibyte committed:
Cover newly hidden/expired elements in existing test

peppy commented on osu.Game.Tests/Visual/Multiplayer/TestSceneMultiSpectatorScreen.cs:
```diff
@@ -90,10 +90,13 @@ public void TestSpectatorPlayerSettingsHidden()
             AddUntilStep("wait for players to load", () => spectatorScreen.AllPlayersLoaded);
-            AddAssert("all player settings hidden", () => this.ChildrenOfType<Player>().All(p => p.ChildrenOfType<PlayerSettingsOverlay>().Single().State.Value == Visibility.Hidden));
+            AddUntilStep("all interactive elements removed", () => this.ChildrenOfType<Player>().All(p =>
+                !p.ChildrenOfType<PlayerSettingsOverlay>().Any() &&
```
Test failure looks related.

frenzibyte replied:
Yeah, should be resolved now.

frenzibyte committed:
Fix failing test and rename to match new behaviour

peppy commented on osu.Game.Tests/Visual/Multiplayer/TestSceneMultiSpectatorScreen.cs:
```diff
@@ -96,7 +96,7 @@ public void TestSpectatorPlayerSettingsHidden()
                 !p.ChildrenOfType<PlayerSettingsOverlay>().Any() &&
                 !p.ChildrenOfType<HoldForMenuButton>().Any() &&
-                !p.ChildrenOfType<SongProgressBar>().Single().ShowHandle));
+                p.ChildrenOfType<SongProgressBar>().SingleOrDefault()?.ShowHandle == false));
```
???

please check the CI again. this is hiding the test rather than correctly testing it. it only failed once out of all the runs. now it's going to silently pass on a wrong condition. i believe you need a wait step as song process is async loaded.

frenzibyte replied:
yeah well, the reason I chose `AddUntilStep` was that it may take more than one frame for this check to pass due to certain components not loading at once.

will go with `AddWaitStep` then with explanatory comment

peppy replied:
what you had was more correct, it just read wrong to me.

frenzibyte replied:
Right... my brian halted while I was pushing that change. I think `AddUntilStep` *with* an explanatory comment should read well.

frenzibyte committed:
Replace until step with wait step with explanatory comment

frenzibyte committed:
Remove arbitrarily-set wait step with until step instead, keeping the comment

frenzibyte committed:
Revert reverted segment to fix failure

bdach commented on osu.Game/Screens/Play/PlayerLoader.cs:
```diff
@@ -48,6 +48,11 @@ public class PlayerLoader : ScreenWithBeatmapBackground
+        /// <summary>
+        /// A fill flow containing the player settings groups, exposed for the ability to hide it from inheritors of the player loader.
+        /// </summary>
+        protected FillFlowContainer<PlayerSettingsGroup> PlayerSettingsGroups;
```
maybe just `PlayerSettings`?

frenzibyte replied:
Sure.

bdach commented on osu.Game.Tests/Visual/Multiplayer/TestSceneMultiSpectatorScreen.cs:
```diff
@@ -80,6 +83,25 @@ public void TestGeneral()
+
+            // components wrapped in skinnable target containers load asynchronously, potentially taking more than one frame to load.
+            // therefore use until step rather than direct assert to account for that.
+            AddUntilStep("all interactive elements removed", () => this.ChildrenOfType<Player>().All(p =>
```
This test step is consistently failing for me in test browser/non-headless.

frenzibyte replied:
It passes locally for me? can you provide the thrown exception and the stack trace? weird it fails

bdach replied:
It just times out after the 50 repeats allowed. No stack trace.

![2021-08-14-190759_232x378_scrot](https://user-images.githubusercontent.com/20418176/129455008-675ce86f-8f35-4ec3-8c79-ec255d035e7b.png)

frenzibyte replied:
Hmm, I'm still not sure what the cause of that could be, as I've tried multiple times with different skins and ended up passing the test. Would need you to debug this failure in that case, as I can't reproduce.

peppy replied:
If anything, I think it will be a client side setting (some kind of HUD toggle or otherwise).

frenzibyte replied:
You're right, turns out hiding the HUD overlay blocks `SkinnableTargetContainer` from fininshing async load of components, as it schedules the lambda and never gets invoked until the HUD overlay gets updated.

Appears to be a trivial one to solve, will do and open another PR for that.

frenzibyte replied:
In addition to https://github.com/ppy/osu/pull/14337, I've forced the HUD visibility mode to "Always" in the test case just to make sense of things.

frenzibyte committed:
PlayerSettingsGroups -> PlayerSettings

frenzibyte committed:
Force HUD visibility mode to "Always" during testing

peppy committed:
Use `PlayerConfiguration` to convey no-seek state

peppy committed:
Avoid new bindable requirement

peppy commented:
I've made some changes to avoid exposing a new bindable from `Player`. Would appreciate a double-check review from someone else.

peppy committed:
Allow seeking via `Player.Seek` even if disabled

peppy committed:
Rename variable to better reflect its purpose

bdach commented on osu.Game/Screens/Play/SongProgress.cs:
```diff
@@ -119,7 +119,8 @@ private void load(OsuColour colours, OsuConfigManager config, DrawableRuleset dr
             if (drawableRuleset != null)
             {
-                AllowSeeking.BindTo(drawableRuleset.HasReplayLoaded);
+                if (player?.Configuration.AllowUserInteraction == true)
+                    ((IBindable<bool>)AllowSeeking).BindTo(drawableRuleset.HasReplayLoaded);
```
This reads a little funny, maybe, but it makes some sense to a degree... I think the biggest throw-off here is `HasReplayLoaded` being bindable. This is really just a set in practice but looks like a strange conditional bind. I don't see how to improve this either, so...

peppy replied:
Can agree; was just leaving it how it was on `master`.

smoogipoo committed:
Privatise setters

smoogipoo committed:
Merge branch 'master' into hide-player-settings-overlay
